File: COA software/backend/app/schemas/train.py
from typing import Optional
from datetime import datetime
from pydantic import BaseModel, ConfigDict

class TrainBase(BaseModel):
    train_number: str
    train_name: str
    train_type: str = "EXPRESS"
    default_direction: str = "UP"
    origin: str = "STN-A"
    destination: str = "STN-C"
    priority: int = 1
    status: str = "SCHEDULED"
    corridor_id: Optional[str] = None

class TrainCreate(TrainBase):
    pass

class TrainUpdate(BaseModel):
    train_name: Optional[str] = None
    train_type: Optional[str] = None
    default_direction: Optional[str] = None
    origin: Optional[str] = None
    destination: Optional[str] = None
    priority: Optional[int] = None

class TrainResponse(TrainBase):
    model_config = ConfigDict(from_attributes=True)
    id: str
    direction: Optional[str] = "UP"
    created_at: datetime
    updated_at: datetime
    is_goods_train: bool = False
    is_passenger_train: bool = True

class TrainScheduleBase(BaseModel):
    train_id: str
    station_id: Optional[str] = None
    corridor_id: Optional[str] = None
    scheduled_date: Optional[datetime] = None
    arrival_time: Optional[datetime] = None
    departure_time: Optional[datetime] = None
    line: Optional[str] = "MAIN_1"
    sequence_number: int = 1

class TrainScheduleCreate(TrainScheduleBase):
    pass

class TrainScheduleResponse(TrainScheduleBase):
    model_config = ConfigDict(from_attributes=True)
    id: str
    created_at: datetime

class TrainMovementBase(BaseModel):
    train_id: str
    corridor_id: Optional[str] = None
    station_id: Optional[str] = None
    event_type: str = "ARRIVAL"
    event_time: Optional[datetime] = None
    actual_time: Optional[datetime] = None
    direction: Optional[str] = "UP"
    line: Optional[str] = "MAIN_1"
    status: str = "SCHEDULED"

class TrainMovementCreate(TrainMovementBase):
    pass

class TrainMovementResponse(TrainMovementBase):
    model_config = ConfigDict(from_attributes=True)
    id: str
    created_at: datetime

File: COA software/frontend/src/types/train.ts
export interface Train {
  id: string
  train_number: string
  train_name: string
  train_type: string
  default_direction: string
  origin: string
  destination: string
  priority: number
  status: string
  corridor_id?: string
  is_goods_train?: boolean
  is_passenger_train?: boolean
  max_speed_kmh?: number
  length_meters?: number
  created_at?: string
}

export interface TrainSchedule {
  id: string
  train_id: string
  train?: Train
  corridor_id: string
  station_id: string
  stop_sequence?: number
  scheduled_date: string
  arrival_time: string
  departure_time: string
  dwell_minutes: number
  halt_duration_minutes?: number
  platform?: string | null
}

export interface TrainMovement {
  id: string
  train_id: string
  train?: Train
  corridor_id: string
  station_id?: string | null
  event_time: string
  event_type: string
  status: string
  delay_minutes: number
  location_km?: number | null
}

export interface GoodsForecast {
  id: string
  corridor_id: string
  forecast_date: string
  hour_start: number
  hour_end: number
  expected_goods_trains: number
  traffic_density: string
  movement_probability: number
  model_version?: string
}

export interface TrainDensity {
  train_count: number
  passenger_count: number
  goods_count: number
  total_movements: number
  traffic_density: string
}

export interface TrainImpact {
  affected_trains: Partial<Train>[]
  affected_train_count: number
  passenger_trains: number
  goods_trains: number
  estimated_delay_minutes: number
  maximum_delay_minutes: number
  impact_level: string
  method: string
}
